desc/equilibrium/coords.py

```diff
@@ -5,14 +5,143 @@
+    """Given coordinates in inbasis, compute corresponding coordinates in outbasis.
+
+    First solves for the computational coordinates that correspond to inbasis, then
+    evaluates outbasis at those locations.
+
+    NOTE: this function cannot be JIT compiled or differentiated with AD.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium to use
+    coords : ndarray, shape(k,3)
+        2D array of input coordinates. Each row is a different
+        coordinate.
+    inbasis, outbasis : tuple of str
+        Labels for input and output coordinates, eg ("R", "phi", "Z") or
+        ("rho", "alpha", "zeta") or any combination thereof. Labels should be the
+        same as the compute function data key
+    tol : float
+        Stopping tolerance.
+    maxiter : int > 0
+        Maximum number of Newton iterations
+    rhomin : float
+        Minimum allowable value of rho (to avoid singularity at rho=0)
+
+
+    """
+    assert (
+        np.isfinite(maxiter) and maxiter > 0
+    ), f"maxiter must be a positive integer, got {maxiter}"
+    assert np.isfinite(tol) and tol > 0, f"tol must be a positive float, got {tol}"
+
+    basis_derivs = [f"{X}_{d}" for X in inbasis for d in ("r", "t", "z")]
+    for key in basis_derivs:
+        assert (
+            key in data_index.keys()
+        ), f"don't have recipe to compute partial derivative {key}"
+
+    # can't use AD or jit compile this because of grid stuff
+    def compute(y, basis):
+        data = eq.compute(basis, grid=Grid(y, sort=False))
+        x = jnp.stack([data[k] for k in basis], axis=-1)
+        return x
+
+    def residual(y):
+        xk = compute(y, inbasis)
+        return xk - coords
+
+    def jac(y):
+        J = compute(y, basis_derivs)
+        J = J.reshape((-1, 3, 3))
+        return J
+
+    @jit
+    def fixup(y):
+        r, t, z = y.T
+        # negative rho -> flip theta
+        t = jnp.where(r < 0, (t + np.pi) % (2 * np.pi), t % (2 * np.pi))
+        r = jnp.abs(r)
+        r = jnp.clip(r, rhomin, 1)
+        z = z % (2 * np.pi)
+        y = jnp.array([r, t, z]).T
+        return y
+
+    # use least squares to avoid singular behavior, cheap for 3x3 systems
+    lstsq = jit(
+        jnp.vectorize(
+            lambda A, b: jnp.linalg.lstsq(A, b)[0], signature="(n,n),(n)->(n)"
+        )
+    )
+
+    # nearest neighbor search on coarse grid for initial guess
+    yg = ConcentricGrid(L=eq.L_grid, M=eq.M_grid, N=int(eq.N_grid * eq.NFP)).nodes
+    xg = compute(yg, inbasis)
+    idx = jnp.zeros(len(coords)).astype(int)
+    coords = jnp.asarray(coords)
+
+    def _distance_body(i, idx):
+        distance = jnp.linalg.norm(coords[i] - xg, axis=-1)
+        k = jnp.argmin(distance)
+        idx = put(idx, i, k)
+        return idx
+
+    idx = fori_loop(0, len(coords), _distance_body, idx)
+    yk = yg[idx]
+    alpha = 0.5
+    yk = fixup(yk)
+    resk = residual(yk)
+
+    for i in range(maxiter):
```

Comment:
Just to make sure I get the algorithm.

You find the initial guesses with a coarse grid nearest neighbor.

Then you are doing a Newton method on $$\mathbf{f}(\mathbf{x}) = \mathbf{yk} - coords$$
where $$\mathbf{x} = (\rho,\theta,\zeta)$$ and  $$\mathbf{x}^{i+1} =  \mathbf{x}^{i} - \mathbf{Df}^{-1}(\mathbf{x}^{i}) @ \mathbf{f}(\mathbf{x^i})$$
then moving $x^i$ over and defining $$\mathbf{d} = \mathbf{x}^{i+1} - \mathbf{x}^i$$ and multiplying by the jacobian yields the least squares system you solve for $\mathbf{d}$:
$$\mathbf{Df}(\mathbf{x}^{i}) \mathbf{d} = -\mathbf{f}(\mathbf{x}^{i})$$

Then you update the guess at the correct coordinates with $\mathbf{d}$ using a backtracking line search.

Did I miss something tho or should your `resk` be negative, based off of my derivation here? 

Or wait, you define $\mathbf{d}$ the opposite of me, so that $\mathbf{x}^{i+1} = \mathbf{x}^i - \mathbf{d}$, ok that clears up my confusion